### DISCUSSION TITLE
`$watch` doesn't destroy and keep reinitializing on a component shown/hidden with `x-if`

### DISCUSSION THREAD

**Discussion by parasshah195** [alpinejs/alpine#3654] [7. Bugs (archived)] [answered]
Hello,

I'm having a strange issue where the watcher does not destroy on `x-if` element hidden, and keeps reinitializing when the component is re-created on condition being true. I'm doing an API query from the component, and this is causing a linearly increasing number of API queries on every destroy and re-init.

A partial and relevant snippet of my setup:

Template:
```
<div x-data>
   <template x-if="$store.filters.is_online !== false">
      <div x-data="filterEventsOnline" x-ref="componentEl">
           {{ content populated from API query here }}
      </div>
   </template>
</div>
```

Component (partial):
```
window.Alpine.data(self.COMPONENT_NAME, function () {
  return {
   // property that holds the API query results
    events: [],
    
    // holds the API query body parameters
    apiBody: self.API_BODY,
    
    init(): void { 
      this.setAPIQueryFilters();
      
      // Initial API query
      this.processQuery();
      
      // setting store watcher to update respective `apiBody` property
      this.$watch(`$store.${FILTER_STORE_NAME}`, () => {
        this.setAPIQueryFilters();
      });
      
      // TODO: fix watcher registering every time extra on the component re-initializing using `x-if`
      this.$watch(
        `apiBody.topics, 
        apiBody.extended_time_available, 
        apiBody.test_date, 
        apiBody.days_of_week`,
        () => {
          // console.log('apiBody watcher triggered');
          // run the API query again and update `events` property
          this.processQuery();
        }
      );
    },

    // Set API body to link to filter store data
    setAPIQueryFilters(): void {
      // const filterStore: FilterAlpineStore = this.$store[FILTER_STORE_NAME];
        
      // reset events
      this.events = [];
        
      // reset pagination
      this.apiBody.start = 1;
        
      this.apiBody.topics = this.$store[FILTER_STORE_NAME].topics;
      this.apiBody.extended_time_available =
      this.$store[FILTER_STORE_NAME].extended_time_available;
      this.apiBody.test_date = this.$store[FILTER_STORE_NAME].test_date;
      this.apiBody.days_of_week = this.$store[FILTER_STORE_NAME].days_of_week;
    },
  };
});
```

**Short summary:** `filters` store is updated by some other events on the page. The whole component is to show and run when the store's `is_online` property is `true`. The component watches for updates on the store (I tried other ways but it didn't update by itself) and updates the `apiBody` property of the component. `apiBody` has a watcher to update query parameters and run the API call again.

**Problem:** The watcher on `apiBody` property never seems to be destroyed when the component hides on the store's `is_online` being `false`. Because when I check network queries and console logged the watcher, new instances of it keep being initialized with the API query count continuing to increase on any change thereafter.

I looked for any other place in code where there could've been a potential leak or multiple calls but that isn't the case. Everything seems to be functioning fine until the instance is re-initialized by `x-if`.

Been scratching my head at this for hours to no avail. Any pointers or suggestions will be helpful. Thank you!

**Comment by ekwoka** [comment, marked as answer]
It actually is.

The issue here is that `$watch` does NOT clean up it's effect.

What you should instead be using is `effect` and `release`

like

```js
Alpine.data('foo', () => {
  const myEffects = [];
  return {
     init() {
       myEffects.push(Alpine.effect(() => doStuffReactively())
    },
    destroy() {
      myEffects.forEach(Alpine.release)
    }
 })
 ```
 
 Basically, watch is not self cleaning, I believe by design but I'm not privy to exactly the reasoning behind this.
 
 Docs could probably be updated to include this information.
 
 One of many reasons I never use `$watch` (or `x-ref` for that matter)

**Reply by parasshah195** [reply]
Thanks for bringing that to light. Didn't know about `Alpine.release`.

Also, does `destroy()` auto-execute (like `init()`) when a component is destroyed? I remember reading in one of the discussion posts here that Alpine does not have any mechanism or function execution when a component is destroyed. Also not mentioned anywhere in the docs.

**Reply by ekwoka** [reply]
It does and has since quite early on in v3.

I believe Caleb hasn't decided that it's stable for some unknown reason.

**Reply by parasshah195** [reply]
In that case, `destroy()` not really safe to use for long-term then, is it?

Also, I don't understand the logic behind `$watch` not cleaning up event listeners bound to a particular component's properties. That's a potential cause for memory leak, because once a component is destroyed, it doesn't make sense to have its watchers still listening to events.

**Reply by ekwoka** [reply]
Well, destroy hasn't changed for a long time.

So, idk personally.

Similar with watch. I don't know. I feel like it came up once, and it was said to be intentional.

You can use an `x-effect` and that would self clean

**Reply by parasshah195** [reply]
Thank you for your help @ekwoka. It'd be nice to hear from @calebporzio about the reasoning on `$watch` not clearing itself on component destroy.
